File: cpp/KalkCpp/Shape/Shape2D/shape2d.cpp
#include "shape2d.h"
#include <string.h>
#include <vector>
#include <math.h>
#include <iostream>



Shape2D::Shape2D(const std::vector<Point2D>& vp2d, const std::string& name, ColorRGB* color): Shape(name, color), points(vp2d)
{

}



bool Shape2D::operator ==(const Shape &s) const
{
    if(Shape::operator ==(s))
    {
        const Shape2D& s2d = static_cast<const Shape2D&>(s);
        if(points == s2d.getPoints())
        {
            return true;
        }
    }
    return false;
}


unsigned int Shape2D::getNumberSide() const
{
    return points.size();
}



unsigned int Shape2D::getNumberVertex() const
{
    return points.size();
}

void Shape2D::translate(const Point2D& p)
{
    for(std::vector<Point2D>::iterator it = points.begin(); it != points.end(); ++it)
    {
        (*it).setX(p.getX() +(*it).getX());
        (*it).setY(p.getY() + (*it).getY());
    }
}

const std::vector<Point2D>& Shape2D::getPoints() const
{
    return points;
}

File: Java/src/Shape/Shape2D/Shape2D.java
package Shape.Shape2D;

import Color.ColorRGB;
import Interface.ICalculable;
import Shape.Point2D;
import Shape.Shape;

import java.util.ArrayList;

public abstract class Shape2D extends Shape implements ICalculable<Shape2D> {
    protected ArrayList<Point2D> points;

    public Shape2D(ArrayList<Point2D> p, String n) {
        super(n);
        points = p;
    }

    public Shape2D(ArrayList<Point2D> p, String name, ColorRGB color) {
        super(name, color);
        points = p;
    }

    public Shape2D(Shape2D s2d) {
        super(s2d);
        points = s2d.points;
    }

    public int getNumberSide() {
        return points.size();
    }

    public abstract double sideToOperator();

    public int getNumberVertex() {
        return points.size();
    }

    public ArrayList<Point2D> getPoints() {
        return points;
    }

    public void translate(Point2D p) {
        for (Point2D point : points) {
            point.setX(p.getX() + point.getX());
            point.setY(p.getY() + point.getY());
        }
    }

    public boolean equalsPoints(ArrayList<Point2D> p) {
        if (points == null || p == null)
            return false;
        if (points.size() != p.size())
            return false;
        for (Point2D points : points) {
            if (!p.contains(points)) {
                return false;
            }
        }
        return true;
    }

    public boolean equals(Object obj) {
        if (super.equals(obj)) {
            Shape2D s2d = (Shape2D) obj;
            return equalsPoints(s2d.points);
        }
        return false;
    }
}
